File: ai-inference/src/ipc_client.cpp
// =============================================================================
// Project: FALX V2
// Lead Architect & Owner: FT-1
// Description: AI engine IPC client (ai-inference/src/ipc_client.cpp).
//              Connects to the falxd Unix socket, sends MapUpdateReq messages
//              when the AI model decides to block an IP, and receives
//              InferRequest batches for analysis.
//
//              Threading model:
//                - recv_thread: reads frames from socket → dispatch
//                - send_thread: drains send_queue → writes frames
//                - heartbeat_thread: sends Heartbeat every 5s
//                All threads share the socket fd (guarded by send_mutex).
// =============================================================================

#include "ipc_client.hpp"
#include "ipc_protocol.hpp"

#include <arpa/inet.h>
#include <atomic>
#include <cerrno>
#include <chrono>
#include <cstring>
#include <iostream>
#include <mutex>
#include <nlohmann/json.hpp>
#include <queue>
#include <sys/socket.h>
#include <sys/un.h>
#include <thread>
#include <unistd.h>

namespace falx::ai {

using json = nlohmann::json;
using namespace falx::ipc;

// ─── Constructor ──────────────────────────────────────────────────────────────
IPCClient::IPCClient(const std::string& socket_path,
                     std::shared_ptr<spdlog::logger> log)
    : socket_path_(socket_path), log_(log), fd_(-1), running_(false) {}

IPCClient::~IPCClient() {
    stop();
}

// ─── Connect ──────────────────────────────────────────────────────────────────
bool IPCClient::connect() {
    fd_ = ::socket(AF_UNIX, SOCK_STREAM, 0);
    if (fd_ < 0) {
        log_->error("socket() failed: {}", std::strerror(errno));
        return false;
    }

    sockaddr_un addr{};
    addr.sun_family = AF_UNIX;
    std::strncpy(addr.sun_path, socket_path_.c_str(), sizeof(addr.sun_path) - 1);

    if (::connect(fd_, reinterpret_cast<sockaddr*>(&addr), sizeof(addr)) < 0) {
        log_->error("connect({}) failed: {}", socket_path_, std::strerror(errno));
        ::close(fd_);
        fd_ = -1;
        return false;
    }

    log_->info("Connected to falxd IPC socket: {}", socket_path_);
    return true;
}

// ─── Start ────────────────────────────────────────────────────────────────────
bool IPCClient::start() {
    if (!connect()) return false;

    running_.store(true);

    recv_thread_ = std::thread([this] { recv_loop(); });
    send_thread_ = std::thread([this] { send_loop(); });
    hb_thread_   = std::thread([this] { heartbeat_loop(); });

    return true;
}

// ─── Stop ─────────────────────────────────────────────────────────────────────
void IPCClient::stop() {
    running_.store(false);
    if (fd_ >= 0) {
        ::shutdown(fd_, SHUT_RDWR);
        ::close(fd_);
        fd_ = -1;
    }
    send_cv_.notify_all();

    if (recv_thread_.joinable()) recv_thread_.join();
    if (send_thread_.joinable()) send_thread_.join();
    if (hb_thread_.joinable())   hb_thread_.join();
}

// ─── Send MapUpdateReq (AI → CP) ─────────────────────────────────────────────
bool IPCClient::send_block_request(
    const std::string& src_ip,
    uint8_t  action,
    uint8_t  threat_score,
    uint8_t  rule_id,
    uint64_t ttl_seconds,
    const std::string& reason
) {
    json payload = {
        {"src_ip",       src_ip},
        {"action",       action},
        {"threat_score", threat_score},
        {"rule_id",      rule_id},
        {"ttl_s",        ttl_seconds},
        {"reason",       reason},
        {"actor",        "ai-engine"},
    };

    auto frame = build_frame(
        MsgType::MapUpdateReq,
        next_seq(),
        payload.dump()
    );

    return enqueue_frame(std::move(frame));
}

// ─── Send Alert (AI → CP) ─────────────────────────────────────────────────────
bool IPCClient::send_alert(
    uint64_t   flow_id,
    const std::string& src_ip,
    const std::string& severity,
    const std::string& threat_type,
    double     confidence,
    const std::string& description
) {
    json payload = {
        {"flow_id",     flow_id},
        {"src_ip",      src_ip},
        {"severity",    severity},
        {"threat_type", threat_type},
        {"confidence",  confidence},
        {"description", description},
    };

    auto frame = build_frame(MsgType::Alert, next_seq(), payload.dump());
    return enqueue_frame(std::move(frame));
}

// ─── Receive Loop ─────────────────────────────────────────────────────────────
void IPCClient::recv_loop() {
    std::vector<uint8_t> buf(HEADER_SIZE + MAX_PAYLOAD_SIZE);

    while (running_.load()) {
        // Read header
        ssize_t n = recv_exact(buf.data(), HEADER_SIZE);
        if (n <= 0) break;

        // Parse header to get payload length
        Frame frame;
        try {
            // Read payload length from bytes 12-15
            uint32_t payload_len =
                static_cast<uint32_t>(buf[12])        |
                (static_cast<uint32_t>(buf[13]) << 8) |
                (static_cast<uint32_t>(buf[14]) << 16)|
                (static_cast<uint32_t>(buf[15]) << 24);

            if (payload_len > MAX_PAYLOAD_SIZE) {
                log_->error("Payload too large: {}", payload_len);
                break;
            }

            // Read payload
            if (payload_len > 0) {
                n = recv_exact(buf.data() + HEADER_SIZE, payload_len);
                if (n <= 0) break;
            }

            frame = parse_frame(buf.data(), HEADER_SIZE + payload_len);
        } catch (const std::exception& e) {
            log_->error("Frame parse error: {}", e.what());
            continue;
        }

        dispatch(frame);
    }

    log_->info("IPC recv loop exiting");
    running_.store(false);
}

// ─── Dispatch Received Frame ──────────────────────────────────────────────────
void IPCClient::dispatch(const Frame& frame) {
    MsgType type = static_cast<MsgType>(frame.header.type);

    switch (type) {
    case MsgType::InferRequest: {
        // CP → AI: packet batch for inference
        std::string json_str(frame.payload.begin(), frame.payload.end());
        try {
            auto req = json::parse(json_str);
            if (infer_request_cb_) {
                infer_request_cb_(req);
            }
        } catch (const std::exception& e) {
            log_->error("InferRequest parse error: {}", e.what());
        }
        break;
    }

    case MsgType::Heartbeat: {
        // Reply with HeartbeatAck using same seq
        auto ack = build_frame(MsgType::HeartbeatAck, frame.header.seq, 0, nullptr, 0);
        enqueue_frame(std::move(ack));
        break;
    }

    case MsgType::HeartbeatAck:
        log_->debug("Heartbeat ACK received (seq={})", frame.header.seq);
        break;

    case MsgType::Error: {
        std::string json_str(frame.payload.begin(), frame.payload.end());
        log_->warn("Error from falxd: {}", json_str);
        break;
    }

    default:
        log_->warn("Unknown msg type: 0x{:02x}", frame.header.type);
    }
}

// ─── Send Loop ────────────────────────────────────────────────────────────────
void IPCClient::send_loop() {
    while (running_.load()) {
        std::unique_lock<std::mutex> lock(send_mutex_);
        send_cv_.wait(lock, [this] {
            return !send_queue_.empty() || !running_.load();
        });

        while (!send_queue_.empty()) {
            auto bytes = serialise(send_queue_.front());
            send_queue_.pop();
            lock.unlock();

            if (!write_all(bytes.data(), bytes.size())) {
                log_->error("Send failed — disconnecting");
                running_.store(false);
                return;
            }
            lock.lock();
        }
    }
}

// ─── Heartbeat Loop ───────────────────────────────────────────────────────────
void IPCClient::heartbeat_loop() {
    while (running_.load()) {
        std::this_thread::sleep_for(std::chrono::seconds(5));
        if (!running_.load()) break;

        auto frame = build_frame(MsgType::Heartbeat, next_seq(), 0, nullptr, 0);
        enqueue_frame(std::move(frame));
    }
}

// ─── Helpers ──────────────────────────────────────────────────────────────────
bool IPCClient::enqueue_frame(Frame&& frame) {
    {
        std::lock_guard<std::mutex> lock(send_mutex_);
        if (send_queue_.size() > 4096) {
            log_->warn("Send queue full — frame dropped");
            return false;
        }
        send_queue_.push(std::move(frame));
    }
    send_cv_.notify_one();
    return true;
}

uint32_t IPCClient::next_seq() {
    return seq_counter_.fetch_add(1, std::memory_order_relaxed);
}

ssize_t IPCClient::recv_exact(uint8_t* buf, size_t n) {
    size_t received = 0;
    while (received < n && running_.load()) {
        ssize_t r = ::recv(fd_, buf + received, n - received, 0);
        if (r <= 0) return r;
        received += r;
    }
    return static_cast<ssize_t>(received);
}

bool IPCClient::write_all(const uint8_t* buf, size_t n) {
    size_t sent = 0;
    while (sent < n) {
        ssize_t w = ::send(fd_, buf + sent, n - sent, MSG_NOSIGNAL);
        if (w <= 0) return false;
        sent += w;
    }
    return true;
}

} // namespace falx::ai

File: ai-inference/ipc_client.py
# =============================================================================
# Project: FALX V2
# Lead Architect & Owner: FT-1
# Description: Python IPC client (ai-inference/ipc_client.py).
#              Python-side implementation of the FLX2 wire protocol.
#              Used when the AI inference layer is implemented in Python
#              (e.g. scikit-learn, PyTorch models) rather than C++.
#
#              Fully async (asyncio) for integration with async ML frameworks.
#              Implements the same protocol as ipc_protocol.hpp and protocol.go.
# =============================================================================

from __future__ import annotations

import asyncio
import json
import logging
import struct
import zlib
from dataclasses import dataclass, field
from enum import IntEnum
from typing import Callable, Awaitable, Optional

log = logging.getLogger("falx.ipc")

# ─── Constants ────────────────────────────────────────────────────────────────
MAGIC_BYTES      = b'\x46\x4C\x58\x32'  # "FLX2"
PROTOCOL_VERSION = 1
HEADER_SIZE      = 20
MAX_PAYLOAD_SIZE = 1 << 20  # 1 MiB

# Header format: magic(4s) version(B) type(B) flags(H) seq(I) len(I) crc32(I)
# All multi-byte fields are LITTLE-ENDIAN
HEADER_FMT    = '<4sBBHIII'
assert struct.calcsize(HEADER_FMT) == HEADER_SIZE


# ─── Message Types ────────────────────────────────────────────────────────────
class MsgType(IntEnum):
    InferRequest  = 0x01
    ConfigPush    = 0x02
    Heartbeat     = 0x03
    InferResponse = 0x81
    MapUpdateReq  = 0x82
    HeartbeatAck  = 0x83
    Alert         = 0x84
    Error         = 0xFF


# ─── Flags ────────────────────────────────────────────────────────────────────
FLAG_ACK_REQUIRED = 1 << 2


# ─── Frame ────────────────────────────────────────────────────────────────────
@dataclass
class Frame:
    msg_type: MsgType
    seq:      int
    flags:    int
    payload:  bytes = field(default=b'')

    @property
    def crc32(self) -> int:
        hdr = self._pack_header(crc32=0)
        return zlib.crc32(hdr[:16] + self.payload) & 0xFFFFFFFF

    def _pack_header(self, crc32: int = 0) -> bytes:
        return struct.pack(
            HEADER_FMT,
            MAGIC_BYTES,
            PROTOCOL_VERSION,
            int(self.msg_type),
            self.flags,
            self.seq,
            len(self.payload),
            crc32,
        )

    def serialise(self) -> bytes:
        crc = self.crc32
        return self._pack_header(crc) + self.payload

    @classmethod
    def from_bytes(cls, hdr_bytes: bytes, payload: bytes) -> 'Frame':
        magic, version, mtype, flags, seq, plen, crc32 = struct.unpack(
            HEADER_FMT, hdr_bytes
        )
        if magic != MAGIC_BYTES:
            raise ValueError(f"Invalid magic: {magic!r}")
        if version != PROTOCOL_VERSION:
            raise ValueError(f"Unsupported protocol version: {version}")

        frame = cls(
            msg_type=MsgType(mtype),
            seq=seq,
            flags=flags,
            payload=payload,
        )
        # Verify CRC
        expected = frame.crc32
        if crc32 != expected:
            raise ValueError(f"CRC mismatch: got 0x{crc32:08x}, expected 0x{expected:08x}")
        return frame


def build_frame(msg_type: MsgType, seq: int, payload_obj: object = None,
                flags: int = 0) -> Frame:
    payload = b''
    if payload_obj is not None:
        payload = json.dumps(payload_obj, separators=(',', ':')).encode()
        flags |= FLAG_ACK_REQUIRED
    return Frame(msg_type=msg_type, seq=seq, flags=flags, payload=payload)


# ─── Async IPC Client ─────────────────────────────────────────────────────────
class AsyncIPCClient:
    """
    Async IPC client for Python-side AI inference workers.
    Connects to falxd's Unix domain socket and handles the FLX2 protocol.
    """

    def __init__(self, socket_path: str = '/var/run/falx/ai.sock') -> None:
        self.socket_path = socket_path
        self._reader: Optional[asyncio.StreamReader]  = None
        self._writer: Optional[asyncio.StreamWriter]  = None
        self._seq     = 0
        self._running = False

        # Callback: called when CP sends an InferRequest
        self.on_infer_request: Optional[Callable[[dict], Awaitable[None]]] = None

    # ── Lifecycle ─────────────────────────────────────────────────────────────
    async def connect(self) -> None:
        self._reader, self._writer = await asyncio.open_unix_connection(
            path=self.socket_path
        )
        self._running = True
        log.info("Connected to falxd IPC socket: %s", self.socket_path)

    async def disconnect(self) -> None:
        self._running = False
        if self._writer:
            self._writer.close()
            await self._writer.wait_closed()

    async def run(self) -> None:
        """Main loop: read frames and dispatch. Call after connect()."""
        hb_task = asyncio.create_task(self._heartbeat_loop())
        try:
            while self._running:
                try:
                    frame = await self._read_frame()
                    await self._dispatch(frame)
                except asyncio.IncompleteReadError:
                    log.info("IPC connection closed by server")
                    break
                except Exception as exc:
                    log.error("Frame error: %s", exc)
        finally:
            hb_task.cancel()

    # ── AI → CP: Block request ────────────────────────────────────────────────
    async def send_block(
        self, src_ip: str, action: int, threat_score: int,
        rule_id: int, ttl_seconds: int, reason: str
    ) -> None:
        payload = {
            'src_ip': src_ip, 'action': action,
            'threat_score': threat_score, 'rule_id': rule_id,
            'ttl_s': ttl_seconds, 'reason': reason,
            'actor': 'ai-engine',
        }
        frame = build_frame(MsgType.MapUpdateReq, self._next_seq(), payload)
        await self._send_frame(frame)
        log.info("Block request sent: %s action=%d score=%d", src_ip, action, threat_score)

    # ── AI → CP: Alert ────────────────────────────────────────────────────────
    async def send_alert(
        self, flow_id: int, src_ip: str, severity: str,
        threat_type: str, confidence: float, description: str
    ) -> None:
        payload = {
            'flow_id': flow_id, 'src_ip': src_ip,
            'severity': severity, 'threat_type': threat_type,
            'confidence': confidence, 'description': description,
        }
        frame = build_frame(MsgType.Alert, self._next_seq(), payload)
        await self._send_frame(frame)

    # ── AI → CP: Inference Response ───────────────────────────────────────────
    async def send_infer_response(self, request_id: int, verdicts: list) -> None:
        payload = {'request_id': request_id, 'verdicts': verdicts}
        frame   = build_frame(MsgType.InferResponse, self._next_seq(), payload)
        await self._send_frame(frame)

    # ── Internal ──────────────────────────────────────────────────────────────
    async def _read_frame(self) -> Frame:
        hdr_bytes = await self._reader.readexactly(HEADER_SIZE)
        _, _, _, _, _, plen, _ = struct.unpack(HEADER_FMT, hdr_bytes)

        if plen > MAX_PAYLOAD_SIZE:
            raise ValueError(f"Payload {plen} exceeds max {MAX_PAYLOAD_SIZE}")

        payload = await self._reader.readexactly(plen) if plen > 0 else b''
        return Frame.from_bytes(hdr_bytes, payload)

    async def _send_frame(self, frame: Frame) -> None:
        self._writer.write(frame.serialise())
        await self._writer.drain()

    async def _dispatch(self, frame: Frame) -> None:
        if frame.msg_type == MsgType.InferRequest:
            if self.on_infer_request and frame.payload:
                req = json.loads(frame.payload)
                await self.on_infer_request(req)

        elif frame.msg_type == MsgType.Heartbeat:
            ack = Frame(MsgType.HeartbeatAck, seq=frame.seq, flags=0)
            await self._send_frame(ack)
            log.debug("Heartbeat → ACK (seq=%d)", frame.seq)

        elif frame.msg_type == MsgType.Error:
            err = json.loads(frame.payload) if frame.payload else {}
            log.warning("Error from falxd: code=%s msg=%s",
                        err.get('code'), err.get('message'))

    async def _heartbeat_loop(self) -> None:
        while self._running:
            await asyncio.sleep(5)
            try:
                frame = Frame(MsgType.Heartbeat, seq=self._next_seq(), flags=0)
                await self._send_frame(frame)
            except Exception:
                break

    def _next_seq(self) -> int:
        self._seq = (self._seq + 1) & 0xFFFFFFFF
        return self._seq


# ─── Action Constants (mirror bpfmaps) ────────────────────────────────────────
class Action(IntEnum):
    Pass       = 1
    Drop       = 2
    Redirect   = 3
    RateLimit  = 4
